File: IOS/Packages/LiquorEngine/Sources/LiquorEngine/PerceivedProof.swift
import Foundation

/// Does it drink like its proof?
///
/// The question behind "do I taste 62.6%": a barrel-proof bourbon that goes
/// down easy is a different bottle from one that scorches at the same strength,
/// and the number on the label does not tell you which you have. Over a
/// collection it is one of the more useful things to know — it is why people
/// keep a cask-strength bottle they can drink neat.
///
/// Heat is recorded on a 1-5 scale because that is the resolution a person
/// actually has. Anything finer would be invented precision.
public enum PerceivedProof: Sendable {

    /// How hot it actually drank.
    public enum Heat: Int, Sendable, Hashable, CaseIterable, Codable {
        case gentle = 1
        case warm = 2
        case firm = 3
        case hot = 4
        case scorching = 5

        public var label: String {
            switch self {
            case .gentle: return "Gentle"
            case .warm: return "Warm"
            case .firm: return "Firm"
            case .hot: return "Hot"
            case .scorching: return "Scorching"
            }
        }
    }

    public enum Verdict: String, Sendable, Hashable, CaseIterable {
        /// Drinks softer than the label says. The compliment.
        case drinksBelowItsProof
        case drinksAtItsProof
        /// Harsher than the strength alone would explain.
        case drinksAboveItsProof

        public var label: String {
            switch self {
            case .drinksBelowItsProof: return "Drinks below its proof"
            case .drinksAtItsProof: return "Drinks about right"
            case .drinksAboveItsProof: return "Drinks above its proof"
            }
        }
    }

    /// What a given strength usually feels like, before the whiskey gets a say.
    ///
    /// Bands rather than a formula: the step from 40% to 45% is barely
    /// noticeable and the step from 60% to 65% is not, so a linear map would be
    /// wrong at both ends.
    public static func expectedHeat(for abv: ABV) -> Heat {
        switch abv.percent {
        case ..<43: return .gentle
        case ..<48: return .warm
        case ..<54: return .firm
        case ..<60: return .hot
        default: return .scorching
        }
    }

    public struct Result: Hashable, Sendable {
        public let abv: ABV
        public let expected: Heat
        public let actual: Heat
        public let verdict: Verdict
        /// Negative means softer than expected.
        public let difference: Int

        public var headline: String { verdict.label }

        public var summary: String {
            let proof = String(format: "%.1f", abv.proof)
            switch verdict {
            case .drinksBelowItsProof:
                return "You would not guess \(proof) proof from tasting it."
            case .drinksAtItsProof:
                return "Tastes about like \(proof) proof should."
            case .drinksAboveItsProof:
                return "Harsher than \(proof) proof alone would explain — worth a "
                    + "few drops of water."
            }
        }
    }

    public static func compare(abv: ABV, felt: Heat) -> Result {
        let expected = expectedHeat(for: abv)
        let difference = felt.rawValue - expected.rawValue

        let verdict: Verdict
        if difference <= -1 {
            verdict = .drinksBelowItsProof
        } else if difference >= 1 {
            verdict = .drinksAboveItsProof
        } else {
            verdict = .drinksAtItsProof
        }

        return Result(
            abv: abv, expected: expected, actual: felt,
            verdict: verdict, difference: difference)
    }
}

File: Android/engine/src/main/kotlin/com/talastack/liquorlog/engine/PerceivedProof.kt
package com.talastack.liquorlog.engine

import java.util.Locale

/**
 * Does it drink like its proof?
 *
 * The question behind "do I taste 62.6%": a barrel-proof bourbon that goes
 * down easy is a different bottle from one that scorches at the same
 * strength, and the number on the label does not tell you which you have.
 * Over a collection it is one of the more useful things to know -- it is why
 * people keep a cask-strength bottle they can drink neat.
 *
 * Heat is recorded on a 1-5 scale because that is the resolution a person
 * actually has. Anything finer would be invented precision.
 */
object PerceivedProof {

    /** How hot it actually drank. */
    enum class Heat(val value: Int, val label: String) {
        GENTLE(1, "Gentle"),
        WARM(2, "Warm"),
        FIRM(3, "Firm"),
        HOT(4, "Hot"),
        SCORCHING(5, "Scorching"),
        ;

        companion object {
            private val byValue: Map<Int, Heat> by lazy { entries.associateBy { it.value } }

            /** The 1-5 a tasting stored, or null for anything off the scale. */
            fun fromValue(value: Int): Heat? = byValue[value]
        }
    }

    enum class Verdict(val storageKey: String, val label: String) {
        /** Drinks softer than the label says. The compliment. */
        DRINKS_BELOW_ITS_PROOF("drinksBelowItsProof", "Drinks below its proof"),
        DRINKS_AT_ITS_PROOF("drinksAtItsProof", "Drinks about right"),
        /** Harsher than the strength alone would explain. */
        DRINKS_ABOVE_ITS_PROOF("drinksAboveItsProof", "Drinks above its proof"),
    }

    /**
     * What a given strength usually feels like, before the whiskey gets a
     * say.
     *
     * Bands rather than a formula: the step from 40% to 45% is barely
     * noticeable and the step from 60% to 65% is not, so a linear map would
     * be wrong at both ends.
     */
    fun expectedHeat(abv: ABV): Heat = when {
        abv.percent < 43 -> Heat.GENTLE
        abv.percent < 48 -> Heat.WARM
        abv.percent < 54 -> Heat.FIRM
        abv.percent < 60 -> Heat.HOT
        else -> Heat.SCORCHING
    }

    data class Result(
        val abv: ABV,
        val expected: Heat,
        val actual: Heat,
        val verdict: Verdict,
        /** Negative means softer than expected. */
        val difference: Int,
    ) {
        val headline: String get() = verdict.label

        val summary: String
            get() {
                // Locale.ROOT: a proof is a number, not a local convention.
                // Without it a comma-decimal locale prints "124,2 proof".
                val proof = String.format(Locale.ROOT, "%.1f", abv.proof)
                return when (verdict) {
                    Verdict.DRINKS_BELOW_ITS_PROOF ->
                        "You would not guess $proof proof from tasting it."
                    Verdict.DRINKS_AT_ITS_PROOF ->
                        "Tastes about like $proof proof should."
                    Verdict.DRINKS_ABOVE_ITS_PROOF ->
                        "Harsher than $proof proof alone would explain — worth a " +
                            "few drops of water."
                }
            }
    }

    fun compare(abv: ABV, felt: Heat): Result {
        val expected = expectedHeat(abv)
        val difference = felt.value - expected.value

        val verdict = when {
            difference <= -1 -> Verdict.DRINKS_BELOW_ITS_PROOF
            difference >= 1 -> Verdict.DRINKS_ABOVE_ITS_PROOF
            else -> Verdict.DRINKS_AT_ITS_PROOF
        }

        return Result(
            abv = abv, expected = expected, actual = felt,
            verdict = verdict, difference = difference,
        )
    }
}
